Mathlib/Topology/Compactness/Lindelof.lean

```diff
@@ -0,0 +1,357 @@
+    IsLindelof s := fun f hf hfs ↦ by
+  contrapose! h
+  simp only [ClusterPt, not_neBot, ← disjoint_iff, SetCoe.forall',
+    (nhds_basis_opens _).disjoint_iff_left] at h
+  choose fsub U hU hUf using h
+  refine ⟨s, U, fun x ↦ (hU x).2, fun x hx ↦ mem_iUnion.2 ⟨⟨x, hx⟩, (hU _).1 ⟩, ?_ ⟩
+  intro t ht h
+  have uinf := f.sets_of_superset (le_principal_iff.1 fsub) h
+  have uninf : ⋂ i ∈ t, (U i)ᶜ ∈ f := by refine (countable_bInter_mem ht).mpr (fun _ _ ↦ hUf _)
+  rw [← compl_iUnion₂] at uninf
+  have uninf := compl_not_mem uninf
+  simp only [compl_compl] at uninf
+  contradiction
+
+/-- A set `s` is Lindelöf if for every family of closed sets whose intersection avoids `s`,
+there exists a countable subfamily whose intersection avoids `s`. -/
+theorem isLindelof_of_countable_subfamily_closed
+    (h : ∀ {ι : Type u} (t : ι → Set X), (∀ i, IsClosed (t i)) → (s ∩ ⋂ i, t i) = ∅ →
+    ∃ u : Set ι, u.Countable ∧ (s ∩ ⋂ i ∈ u, t i) = ∅) :
+    IsLindelof s :=
+  isLindelof_of_countable_subcover fun U hUo hsU ↦ by
+    rw [← disjoint_compl_right_iff_subset, compl_iUnion, disjoint_iff] at hsU
+    rcases h (fun i ↦ (U i)ᶜ) (fun i ↦ (hUo _).isClosed_compl) hsU with ⟨t, ht⟩
+    refine ⟨t, ?_⟩
+    rwa [← disjoint_compl_right_iff_subset, compl_iUnion₂, disjoint_iff]
+
+/-- A set `s` is Lindelöf if and only if
+for every open cover of `s`, there exists a countable subcover. -/
+theorem isLindelof_iff_countable_subcover :
+    IsLindelof s ↔ ∀ {ι : Type u} (U : ι → Set X),
+      (∀ i, IsOpen (U i)) → (s ⊆ ⋃ i, U i) → ∃ t : Set ι, t.Countable ∧ s ⊆ ⋃ i ∈ t, U i :=
+  ⟨fun hs ↦ hs.elim_countable_subcover, isLindelof_of_countable_subcover⟩
+
+/-- A set `s` is Lindelöf if and only if
+for every family of closed sets whose intersection avoids `s`,
+there exists a countable subfamily whose intersection avoids `s`. -/
+theorem isLindelof_iff_countable_subfamily_closed :
+    IsLindelof s ↔ ∀ {ι : Type u} (t : ι → Set X),
+    (∀ i, IsClosed (t i)) → (s ∩ ⋂ i, t i) = ∅
+    → ∃ u : Set ι, u.Countable ∧ (s ∩ ⋂ i ∈ u, t i) = ∅ :=
+  ⟨fun hs ↦ hs.elim_countable_subfamily_closed, isLindelof_of_countable_subfamily_closed⟩
+
+/-- The empty set is a Lindelof set. -/
+@[simp]
+theorem isLindelof_empty : IsLindelof (∅ : Set X) := fun _f hnf _ hsf↦
+  Not.elim hnf.ne <| empty_mem_iff_bot.1 <| le_principal_iff.1 hsf
+
+/-- A singleton set is a Lindelof set. -/
+@[simp]
+theorem isLindelof_singleton {x : X} : IsLindelof ({x} : Set X) := fun f hf _ hfa ↦
+  ⟨x, rfl, ClusterPt.of_le_nhds'
+    (hfa.trans <| by simpa only [principal_singleton] using pure_le_nhds x) hf⟩
+
+theorem Set.Subsingleton.isLindelof (hs : s.Subsingleton) : IsLindelof s :=
+  Subsingleton.induction_on hs isLindelof_empty fun _ ↦ isLindelof_singleton
+
+/-- X is a Lindelöf space iff every open cover has a countable subcover.-/
+class LindelofSpace (X : Type*) [TopologicalSpace X] : Prop where
+  /-- In a Lindelöf space, `Set.univ` is a Lindelöf set. -/
+  isLindelof_univ : IsLindelof (univ : Set X)
```

Comment:
One more typeclass I added in my PR is `StronglyLindelofSpace` assuming that every open set is Lindelof and implying that every set is Lindelof.

Reply:
Yes, I saw that one, thought it'd be a good follow up PR to add the API for Strongly/Hereditarily Lindelöf spaces! Would you like them in this PR already?